cassiopeia/solver/graph_utilities.py

```diff
@@ -0,0 +1,197 @@
+    ip = {}
+    new_S = set(S)
+    total_weight = 2 * sum([G[e[0]][e[1]]["weight"] for e in G.edges()])
+    num = sum(
+        [G[e[0]][e[1]]["weight"] for e in G.edges() if check_if_cut(e[0], e[1], new_S)]
+    )
+    denom = sum([sum([G[u][v]["weight"] for v in G.neighbors(u)]) for u in new_S])
+    if num == 0:
+        return list(new_S)
+    # curr_score = num/min(denom, total_weight-denom)
+
+    def set_ip(u):
+        if min(denom + delta_d[u], total_weight - denom - delta_d[u]) == 0:
+            ip[u] = 1000
+        else:
+            ip[u] = (num + delta_n[u]) / min(
+                denom + delta_d[u], total_weight - denom - delta_d[u]
+            ) - num / min(denom, total_weight - denom)
+
+    for u in G.nodes():
+        d = sum([G[u][v]["weight"] for v in G.neighbors(u)])
+        if d == 0:
+            return [u]
+        c = sum(
+            [G[u][v]["weight"] for v in G.neighbors(u) if check_if_cut(u, v, new_S)]
+        )
+        delta_n[u] = d - 2 * c
+        if u in new_S:
+            delta_d[u] = -d
+        else:
+            delta_d[u] = d
+        set_ip(u)
+
+    # TODO
+    all_neg = False
+    iters = 0
+
+    while (not all_neg) and (iters < len(G.nodes)):
+        best_potential = 0
+        best_index = None
+        for v in G.nodes():
+            if ip[v] < best_potential:
+                best_potential = ip[v]
+                best_index = v
+        if not best_index is None:
+            num += delta_n[best_index]
+            denom += delta_d[best_index]
+            for j in G.neighbors(best_index):
+                if check_if_cut(best_index, j, new_S):
+                    delta_n[j] += 2 * G[best_index][j]["weight"]
+                else:
+                    delta_n[j] -= 2 * G[best_index][j]["weight"]
+                set_ip(j)
+            delta_n[best_index] = -delta_n[best_index]
+            delta_d[best_index] = -delta_d[best_index]
+            set_ip(best_index)
+            if best_index in new_S:
+                new_S.remove(best_index)
+            else:
+                new_S.add(best_index)
+            # print("curr scores:", num/min(denom, total_weight - denom))
+        else:
+            all_neg = True
+        iters += 1
+    if display:
+        print("sgreed+ score, ", num / min(denom, total_weight - denom))
+    return list(new_S)
+
+
+def similarity(u, v, cm, missing_char, w=None):
+    # TODO Optimize this using masks
+    k = cm.shape[0]
+    if w is None:
+        return sum(
+            [
+                1
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
+    else:
+        return sum(
+            [
+                w[i][cm.iloc[u, i]]
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
```

Comment:
Let's save all of this for a future PR where we check in the graph-based greedy code.